[tutorials] build | svg tutorial (#736)
lit/lit.dev: packages/lit-dev-content/site/tutorials/content/svg-templates/03.md

clip-path above. As a result, tjhe 300px square visually appears to be
a 200px black square.

```TS
const helloTile = svg`
	<rect
		clip-path="url(#rect-clip)"
		width="300"
		height="300"
		fill="#000">
	</rect>
```

### Apply

This demo uses a `<clipPath>` to create a tile containing five groups
of rotated text.

To accomplish this, first, create a clip path composed of a 200px
square.

```TS
const createClipPath = () => svg`
	<clipPath id="rect-clip">
		<rect width="200" height="200"></rect>
	</clipPath>
`;
```

After, move the _motif_ inside `<defs>` and give it an id called
`chars-rotated`.

```TS
@customElement('repeat-pattern')
export class RepeatPattern extends LitElement {

	...

	render() {
		return html`
			<svg>
				<defs>
					${createClipPath()}
					${createElement(this.chars)}
					${createMotif(
						this.numPrints,
						this.rotationOffset,
					)}
				</defs>
			</svg>
    	`;
	}
}
```

Now that the motif is available in `<defs>`, make a function called
`createTile` that returns a group of rotated texts via `<use>`.

The placement and count is arbitrary. Try different variations! The
demo uses five motifs to create an offset clipped along the x axis.

> ```TS

Reviewer: Could this leverage lit templating to make it easy to edit one spot to generate many?

Author: i think so, this spot is where the "tiles" are created. So anything done here is the "repeat" part of the pattern

This section of the tutorial covers the basics of `<clipPath>` and
`<g>` elements in SVG including how to:

- Create a clip path
- Create a group of groups with `<use>`
- Apply a clip path to a group

### Learn

In pattern making, a _tile_ is a polygon that can be repeated across
a two dimensional plane. The tile in this demo will be a series of
transformed texts.

Clip paths are polygons used to restrict where elements are "painted".

> In SVG, a clip path is defined with the `<clipPath>` element.

Reviewer: ```suggestion
In SVG, a clip path is defined with the [`<clipPath>` element](https://developer.mozilla.org/en-US/docs/Web/CSS/clip-path).
```

Author: sounds good! done

rectangle with an `id` of `rect-clip`.

```ts
const helloClipPath = svg`
	<clipPath>
		<rect width="200" height="200"></rect>
	</clipPath>
`;
```

In order to apply a clip path to an SVG element, the `clip-path`
property should reference a `<clipPath>` element by an `id`. The
`<rect>` below will be clipped by the clip-path above by referencing
`#react-clip`.

```ts
const helloTile = svg`
	<rect
		clip-path="url(#rect-clip)"
		width="300"
		height="300"
		fill="#000">
	</rect>
```

### Apply

Create a clip path composed of a 200px square. This `<clipPath>` will
restrict paints to create a _tile_. 

```ts
const createClipPath = () => svg`
	<clipPath id="rect-clip">
		<rect width="200" height="200"></rect>
	</clipPath>
`;
```

Then give the motif an `id` called `chars-rotated`.

```ts
const createMotif = (
	numPrints: number,
	offset: number = 0,
): SVGTemplateResult => {

	...

	return svg`
		<g
			id="chars-rotated"
			transform="translate(50, 50)">
				${prints}
		</g>`;
}
```

After, move the _motif_ inside `<defs>`.

```ts
@customElement('repeat-pattern')
export class RepeatPattern extends LitElement {

	...

	render() {
		return html`
			<svg>
				<defs>

> 					${createClipPath()}

Reviewer: you never mention this line

Author: I think this one changed but don't know to what. prettyu sure this chapter reads better though

					${createElement(this.chars)}
					${createMotif(
						this.numPrints,
						this.rotationOffset,
					)}
				</defs>
			</svg>
    	`;
	}
}
```

Now that the motif is available in `<defs>`, make a function called

> `createTile` that returns a group of rotated texts via `<use>`.

Reviewer: You never say to use it

Author: changed the `#chars-rotated` id to `#motif` to make it more clear

  numPrints,
  offset = 0,
) => {
  ...

  return svg`
    <g
      id="motif"
      transform="translate(50, 50)">
        ${prints}
    </g>
  `;
}
```

{% endswitchable-sample %}

After, move the `createMotif` inside `<defs>`.

```ts
export class RepeatPattern extends LitElement {
  ...

  render() {
    return html`
      <svg height="100%" width="100%">
        <defs>
          ${createTileBoundary()}
          ${createElement(this.chars)}
          ${createMotif(
            this.numPrints,
            this.rotationOffset,
          )}
        </defs>
      </svg>
    `;
  }
}
```

Make a function called `createTile`. This function will utilize the motif
and clipPath now available in `<defs>` by referencing `#motif` in a
`<use>` element, and `url(#rect-clip)` in the `clip-path` attribute.

```ts
const createTile = () => svg`
  <g clip-path="url(#rect-clip)">
    <use transform="translate(0, 0)" href="#motif"></use>
    <use transform="translate(0, 100)" href="#motif"></use>
    <use transform="translate(100, -50)" href="#motif"></use>
    <use transform="translate(100, 50)" href="#motif"></use>
    <use transform="translate(100, 150)" href="#motif"></use>
  </g>
`;
```

> This demo composes a tile containing five groups of rotated text to create

Reviewer 2: Need the "put `createTile()`" in the render function instruction.

Author: sounds good! done!

This section of the tutorial covers the basics of `<clipPath>` and
`<g>` elements in SVG including how to:

- Create a clip path
- Create a group of groups with `<use>`
- Apply a clip path to a group

### Learn

In pattern making, a _tile_ is a polygon that can be repeated across
a two dimensional plane. The tile in this demo will be a series of
transformed texts.

Clip paths are polygons used to restrict where elements are "painted".
In SVG, a clip path is defined with the [`<clipPath>`](https://developer.mozilla.org/en-US/docs/Web/CSS/clip-path)
element.

In the example below, a clip path will be created in the shape of a 

> rectangle with an `id` of `rect-clip`.

Reviewer: You forgot the `id` in the sample

Author: done!

```ts
const helloClipPath = svg`
	<clipPath>
		<rect width="200" height="200"></rect>
	</clipPath>
`;
```

In order to apply a clip path to an SVG element, the `clip-path`
property should reference a `<clipPath>` element by an `id`. The
`<rect>` below will be clipped by the clip-path above by referencing
`#react-clip`.

```ts
const helloTile = svg`
	<rect
		clip-path="url(#rect-clip)"
		width="300"
		height="300"
		fill="#000">
	</rect>
```

### Apply

> Create a clip path composed of a 200px square. This `<clipPath>` will
> restrict paints to create a _tile_. Give the clip-path an `id` called `rect-clip`.

Reviewer: ```suggestion
Use `<clipPath>` to create a 200px square _tile_. This `<clipPath>` will
restrict anything painted inside of it to its 200px square boundaries. Give
the clip-path an `id` called `rect-clip`.
```

Author: done!

```ts

> const createClipPath = () => svg`

Reviewer: Sorry this is late in revew, but why don't you refer to this as its concept like `createMotif` and `createElement`? Why not `createTileBoundary` or something of the like?

Author: lets do that!

Author: done!

	<clipPath id="rect-clip">
		<rect width="200" height="200"></rect>
	</clipPath>
`;
```

Then give the `<g>` returned by `createMotif` an `id` called `motif`.

{% switchable-sample %}

```ts
const createMotif = (
	numPrints: number,
	offset: number = 0,
): SVGTemplateResult => {
	...

	return svg`
		<g
			id="motif"
			transform="translate(50, 50)">
				${prints}
		</g>
  `;
}
```

```js
const createMotif = (
	numPrints,
	offset = 0,
) => {
	...

	return svg`
		<g
			id="motif"
			transform="translate(50, 50)">
				${prints}
		</g>
  `;
}
```

{% endswitchable-sample %}

After, move the `createMotif` inside `<defs>`.

{% switchable-sample %}

```ts

> @customElement('repeat-pattern')

Reviewer: if you remove the calls to customElement, both the TS and JS are the same and you don't have to make a switchable sample here

export class RepeatPattern extends LitElement {
	...

	render() {
		return html`
			<svg height="100%" width="100%">
				<defs>
					${createClipPath()}
					${createElement(this.chars)}
					${createMotif(
						this.numPrints,
						this.rotationOffset,
					)}
				</defs>
			</svg>
    `;
	}
}
```

```js
export class RepeatPattern extends LitElement {
	...

  render() {
    return html`
			<svg height="100%" width="100%">
				<defs>
					${createClipPath()}
					${createElement(this.chars)}
					${createMotif(this.numPrints, this.rotationOffset)}
				</defs>

>         		${createTile()}

Reviewer: This references `createTile` before we define it

Author: done!

			</svg>
    `;
  }
}
customElements.define('repeat-pattern', RepeatPattern);
```
{% endswitchable-sample %}

> Make a function called `createTile`. This function will utilize the motif
> now available in `<defs>` by referencing `#motif` in a `<use>` element.

Reviewer: ```suggestion
Make a function called `createTile`. This function will utilize the motif
and clipPath now available in `<defs>` by referencing `#motif` in a
`<use>` element, and `url(#rect-clip)` in the `clip-path` attribute.
```

Author: Done!

### Overview

This is chapter of the demo covers the basics of `<defs>` and `<use>` in
SVG including how to:

- define a group in defs
- create a group of groups with `<use>`
- create clip path
- apply clip paths to a group

### Learn

The tile in this demo will be a series of transformed texts.

> Clip paths can be used to corral a group of visual elements. In SVG

Reviewer: corral I don't feel like truly explains what it does. Maybe pull wording from MDN?

https://developer.mozilla.org/en-US/docs/Web/SVG/Element/clipPath

Author: done!

> Clip paths can be used to corral a group of visual elements. In SVG

Reviewer: Add comma after SVG

Author: i think that disappeared

a clip path is defined with the `<clipPath>` element. It will take the
shape of its child elements.

In the example below, a clip path will display contents within a 
rectangle with an id of `rect-clip`.

```TS
const helloClipPath = svg`
	<clipPath id="rect-clip">
		<rect width="200" height="200"></rect>
	</clipPath>
`;
```

> In pattern making, a _tile_ is a polygon that can be repeated across

Author: done!

> a two dimensional plane. This demo uses a `<clip-path>` to create a

Reviewer: `<clip-path>` -> `<clipPath>`

Author: done!

tile containing five groups of rotated text.

In the example below a 300px black square will be clipped into visually

> looking like a 200px black square.

Reviewer: This is hard to visualize without an actual example. Perhaps you can merge the learn stuff in between instructions in apply?

Author: changed the wording a bit, i think this is better visualized in the demo

```TS
const helloTile = svg`
	<g clip-path="url(#rect-clip)">
		<rect width="300" height="300" fill="#000"></react>
	</g>`;
```

### Apply

> First, create a clip path composed of a 200px square.

Reviewer: make sure you mention this is to create the tile

Author: done!

```TS
const createClipPath = () => svg`
	<clipPath id="rect-clip">
		<rect width="200" height="200"></rect>
	</clipPath>
`;
```

After, move the _motif_ inside the `<defs>` element and give it an
id called `chars-rotated`.

> ```TS

Reviewer: In this example I don't see `chars-rotated`. Not quite sure where I should be adding it

Author: made the example more explicit!
